include/wil/coroutine.h

```diff
@@ -0,0 +1,791 @@
+            WI_ASSERT((previous == running_ptr()) || (previous == completed_ptr()));
+            return previous == running_ptr();
+        }
+
+        T client_await_resume()
+        {
+            return m_holder.get_value();
+        }
+    };
+
+    template<typename T>
+    struct task_promise : promise_base<T>
+    {
+        template<typename U>
+        void return_value(U&& value)
+        {
+            this->set_value(wistd::forward<U>(value));
+        }
+
+        template<typename Dummy = void>
+        wistd::enable_if_t<!wistd::is_reference_v<T>, Dummy>
+            return_value(T const& value)
+        {
+            this->set_value(value);
+        }
+    };
+
+    template<>
+    struct task_promise<void> : promise_base<void>
+    {
+        void return_void()
+        {
+            this->set_value();
+        }
+    };
+
+    template<typename T>
+    struct promise_deleter
+    {
+        void operator()(promise_base<T>* promise) const noexcept
+        {
+            promise->abandon();
+        }
+    };
+
+    template<typename T>
+    using promise_ptr = wistd::unique_ptr<promise_base<T>, promise_deleter<T>>;
+
+    template<typename T>
+    struct agile_awaiter
+    {
+        agile_awaiter(promise_ptr<T>&& initial) : promise(wistd::move(initial)) { }
+
+        promise_ptr<T> promise;
+
+        bool await_ready()
+        {
+            return promise->client_await_ready();
+        }
+
+        auto await_suspend(__WI_COROUTINE_NAMESPACE::coroutine_handle<> handle)
+        {
+            // Use the default resumer.
+            return promise->client_await_suspend(handle.address(), nullptr);
+        }
+
+        T await_resume()
+        {
+            return promise->client_await_resume();
+        }
+    };
+
+    template<typename T>
+    struct task_base
+    {
+        auto resume_any_thread() && noexcept
+        {
+            return agile_awaiter<T>{ wistd::move(promise) };
+        }
+
+        // You must #include <ole2.h> before <wil\coroutine.h> to enable apartment-aware awaiting.
+        auto resume_same_apartment() && noexcept;
+
+        // Compiler error message metaprogramming: Tell people that they
+        // need to use std::move() if they try to co_await an lvalue.
+        struct cannot_await_lvalue_use_std_move { void await_ready() {} };
+        cannot_await_lvalue_use_std_move operator co_await() & = delete;
+
+        // You must #include <synchapi.h> (usually via <windows.h>) to enable synchronous waiting.
+        decltype(auto) get() &&;
+
+    protected:
+        task_base(task_promise<T>* initial = nullptr) noexcept : promise(initial) {}
+
+        template<typename D>
+        D& assign(D* self, task_base&& other) noexcept
+        {
+            static_cast<task_base&>(*this) = wistd::move(other);
+            return *self;
+        }
+
+    private:
+        promise_ptr<T> promise;

```

Comment:
As part o trying this on TTD we found a need to query a task for whether it's `empty()` (whether the `promise` pointer is `nullptr`. In a way, a task object has two states: it either contains a coroutine that will produce a value or it doesn't. Default-constructed task objects, and task objects that have been awaited (via `co_await` or via `get()`) don't contain a coroutine, so in STL container parlance they are `empty()`.

An alternative is to implement the explicit `operator bool()` like the STL does for smart pointers and for `optional`, but for a task object there are two possible meanings of the bool conversion. A boolean of `true` could mean `!empty()` or it could mean that the coroutine has already produced a value. This semantic ambiguity makes me prefer an explicitly-named function.

Aside: providing another function to check for coroutine completion would also be simple and might help in some scenarios.